File: packages/core/src/core/telemetry.ts
/**
 * Telemetry manager with OpenTelemetry integration
 * @module core/telemetry
 */

import * as otel from '@opentelemetry/api';

/**
 * Span attributes specific to security operations
 */
export interface SecuritySpanAttributes {
  /** MITRE ATT&CK technique ID (e.g., T1059) */
  mitreTechnique?: string;
  /** MITRE ATT&CK tactic name */
  mitreTactic?: string;
  /** Security event type */
  securityEventType?: string;
  /** Threat severity */
  threatSeverity?: string;
  /** Actor identifier */
  actorId?: string;
  /** Resource identifier */
  resourceId?: string;
  /** Action performed */
  action?: string;
  /** Result of the action */
  result?: string;
}

/**
 * Span creation options
 */
export interface SpanOptions {
  /** Span name */
  name: string;
  /** Span kind */
  kind?: otel.SpanKind;
  /** Span attributes */
  attributes?: Record<string, string | number | boolean>;
  /** Parent span context */
  parentContext?: otel.Context;
  /** Security-specific attributes */
  security?: SecuritySpanAttributes;
}

/**
 * Telemetry manager for distributed tracing
 */
export class TelemetryManager {
  /** OpenTelemetry tracer instance */
  private tracer: otel.Tracer;
  /** Whether telemetry is enabled */
  private enabled: boolean;

  /**
   * Create a new TelemetryManager
   * @param serviceName - OpenTelemetry service name
   * @param serviceVersion - Service version
   * @param enabled - Whether telemetry is enabled
   */
  constructor(serviceName: string = 'msf-framework', serviceVersion: string = '1.0.0', enabled: boolean = true) {
    this.tracer = otel.trace.getTracer(serviceName, serviceVersion);
    this.enabled = enabled;
  }

  /**
   * Start a new span with context management
   * @param options - Span creation options
   * @returns The created span
   * @example
   * ```typescript
   * const span = telemetry.startSpan({ name: 'process-request', kind: SpanKind.SERVER });
   * try {
   *   // do work
   *   span.end();
   * } catch (err) {
   *   span.recordException(err);
   *   span.setStatus({ code: SpanStatusCode.ERROR });
   *   throw err;
   * }
   * ```
   */
  startSpan(options: SpanOptions): otel.Span {
    if (!this.enabled) {
      return new NoopSpan();
    }

    const attributes: Record<string, string | number | boolean> = { ...options.attributes };
    if (options.security) {
      const sec = options.security;
      if (sec.mitreTechnique) attributes['security.mitre.technique'] = sec.mitreTechnique;
      if (sec.mitreTactic) attributes['security.mitre.tactic'] = sec.mitreTactic;
      if (sec.securityEventType) attributes['security.event.type'] = sec.securityEventType;
      if (sec.threatSeverity) attributes['security.threat.severity'] = sec.threatSeverity;
      if (sec.actorId) attributes['security.actor.id'] = sec.actorId;
      if (sec.resourceId) attributes['security.resource.id'] = sec.resourceId;
      if (sec.action) attributes['security.action'] = sec.action;
      if (sec.result) attributes['security.result'] = sec.result;
    }

    const spanOptions: otel.SpanOptions = {
      kind: options.kind,
      attributes,
    };

    const parentContext = options.parentContext;
    if (parentContext) {
      return this.tracer.startSpan(options.name, spanOptions, parentContext);
    }
    return this.tracer.startSpan(options.name, spanOptions);
  }

  /**
   * Create a security-focused span with MITRE ATT&CK tagging
   * @param name - Span name
   * @param mitreTechnique - MITRE ATT&CK technique ID
   * @param attributes - Additional attributes
   * @returns The created span
   * @example
   * ```typescript
   * const span = telemetry.createSecuritySpan('detect-injection', 'T1059.001', {
   *   severity: 'high',
   *   action: 'scan',
   * });
   * ```
   */
  createSecuritySpan(
    name: string,
    mitreTechnique: string,
    attributes: Record<string, string | number | boolean> = {}
  ): otel.Span {
    return this.startSpan({
      name,
      kind: otel.SpanKind.INTERNAL,
      attributes,
      security: {
        mitreTechnique,
        securityEventType: name,
      },
    });
  }

  /**
   * Execute a function within a span context
   * @param options - Span options
   * @param fn - Function to execute
   * @returns Function result
   * @example
   * ```typescript
   * const result = await telemetry.withSpan({ name: 'auth' }, async (span) => {
   *   return await authenticate(credentials);
   * });
   * ```
   */
  async withSpan<T>(options: SpanOptions, fn: (span: otel.Span) => Promise<T>): Promise<T> {
    const span = this.startSpan(options);
    try {
      const result = await fn(span);
      span.setStatus({ code: otel.SpanStatusCode.OK });
      return result;
    } catch (error) {
      span.recordException(error as Error);
      span.setStatus({ code: otel.SpanStatusCode.ERROR, message: (error as Error).message });
      throw error;
    } finally {
      span.end();
    }
  }

  /**
   * Get the current active span
   * @returns Current span or undefined
   */
  getActiveSpan(): otel.Span | undefined {
    const span = otel.trace.getActiveSpan();
    return span && span.spanContext().traceId ? span : undefined;
  }

  /**
   * Get the current trace context for propagation
   * @returns Trace context headers
   * @example
   * ```typescript
   * const headers = telemetry.getTraceContext();
   * fetch('http://api', { headers });
   * ```
   */
  getTraceContext(): Record<string, string> {
    const headers: Record<string, string> = {};
    const propagator = otel.propagation;
    propagator.inject(otel.context.active(), headers, {
      set(carrier, key, value) {
        carrier[key] = value as string;
      },
    });
    return headers;
  }

  /**
   * Enable or disable telemetry
   * @param enabled - Whether telemetry should be enabled
   */
  setEnabled(enabled: boolean): void {
    this.enabled = enabled;
  }

  /**
   * Check if telemetry is enabled
   * @returns True if enabled
   */
  isEnabled(): boolean {
    return this.enabled;
  }
}

/**
 * Noop span implementation for when telemetry is disabled
 */
class NoopSpan implements otel.Span {
  spanContext(): otel.SpanContext {
    return { traceId: '00000000000000000000000000000000', spanId: '0000000000000000', traceFlags: 0 };
  }
  updateName(): this { return this; }
  setAttribute(): this { return this; }
  setAttributes(): this { return this; }
  addEvent(): this { return this; }
  setStatus(): this { return this; }
  end(): void {}
  isRecording(): boolean { return false; }
  recordException(): void {}
}

let _telemetry: TelemetryManager | null = null;

/**
 * Get the global telemetry manager singleton
 * @param serviceName - Service name (only on first call)
 * @param serviceVersion - Service version (only on first call)
 * @param enabled - Whether telemetry is enabled (only on first call)
 * @returns TelemetryManager instance
 * @example
 * ```typescript
   const telemetry = getTelemetry();
   * const span = telemetry.startSpan({ name: 'my-operation' });
   * ```
   */
export function getTelemetry(
  serviceName?: string,
  serviceVersion?: string,
  enabled?: boolean
): TelemetryManager {
  if (!_telemetry) {
    _telemetry = new TelemetryManager(serviceName, serviceVersion, enabled);
  }
  return _telemetry;
}

/**
 * Convenience function to create a span
 * @param name - Span name
 * @param attributes - Optional attributes
 * @returns The created span
 * @example
 * ```typescript
   const span = createSpan('process-request');
   * span.end();
   * ```
   */
export function createSpan(
  name: string,
  attributes: Record<string, string | number | boolean> = {}
): otel.Span {
  return getTelemetry().startSpan({ name, attributes });
}

File: master_security_python/src/master_security/core/telemetry.py
"""
Master Security Framework - OpenTelemetry Tracing Engine
=========================================================

Native OpenTelemetry integration for distributed tracing
of all security operations.

Features:
    - Automatic span creation for security operations
    - Security context propagation
    - Error recording with security attributes
    - Correlation with structured logs
    - MITRE ATT&CK technique tagging

Usage:
    >>> from master_security.core import get_telemetry, create_span
    >>> telemetry = get_telemetry()
    >>> with create_span("validate_jwt") as span:
    ...     result = validate_jwt(token)
"""

from __future__ import annotations

import os

import secrets
from contextlib import contextmanager
from typing import Any, Generator, Optional

from opentelemetry import trace, context
from opentelemetry.trace import Span, Status, StatusCode, TracerProvider
from opentelemetry.sdk.trace import TracerProvider as SDKTracerProvider
from opentelemetry.sdk.trace.export import (
    BatchSpanProcessor,
    SpanExporter,
    ConsoleSpanExporter,
)
from opentelemetry.sdk.resources import Resource

from master_security.core.logger import get_logger

logger = get_logger("msf.telemetry")

MSF_SERVICE_NAME = "master-security-framework"
MSF_SERVICE_VERSION = "1.0.0"


class TelemetryManager:
    """
    OpenTelemetry manager for MSF.

    Provides tracer instances and span utilities for
    all security operations.

    Attributes:
        service_name: Service name for telemetry
        endpoint: OTLP exporter endpoint
        enabled: Whether telemetry is enabled

    Example:
        >>> tm = TelemetryManager()
        >>> with tm.start_span("auth.validate_jwt") as span:
        ...     span.set_attribute("user.id", "u123")
    """

    def __init__(
        self,
        service_name: str = MSF_SERVICE_NAME,
        endpoint: str = "http://localhost:4318",
        enabled: bool = None,
    ) -> None:
        self.service_name = service_name
        self.endpoint = endpoint
        self.enabled = enabled if enabled is not None else os.getenv("MSF_OTEL_ENABLED", "true").lower() != "false"
        self._tracer = trace.get_tracer(service_name, MSF_SERVICE_VERSION)

        if enabled:
            resource = Resource.create({
                "service.name": service_name,
                "service.version": MSF_SERVICE_VERSION,
                "telemetry.sdk.name": "msf",
                "telemetry.sdk.language": "python",
            })
            provider = SDKTracerProvider(resource=resource)
            provider.add_span_processor(
                BatchSpanProcessor(ConsoleSpanExporter())
            )
            trace.set_tracer_provider(provider)
            self._tracer = trace.get_tracer(service_name, MSF_SERVICE_VERSION)
            logger.info("msf.telemetry.initialized", endpoint=endpoint)

    @contextmanager
    def start_span(
        self,
        name: str,
        attributes: Optional[dict[str, Any]] = None,
    ) -> Generator[Span, None, None]:
        """
        Start a new security operation span.

        Args:
            name: Span name (e.g., "auth.validate_jwt")
            attributes: Optional span attributes

        Yields:
            Active Span object.
        """
        if not self.enabled:
            class _DummySpan:
                def set_attribute(self, *a, **kw): pass
                def set_status(self, *a, **kw): pass
                def record_exception(self, *a, **kw): pass
                def is_recording(self): return False
                def add_event(self, *a, **kw): pass
                def end(self): pass
            yield _DummySpan()
            return

        with self._tracer.start_as_current_span(name) as span:
            span.set_attribute("msf.operation", name)
            span.set_attribute("msf.version", MSF_SERVICE_VERSION)
            if attributes:
                for key, value in attributes.items():
                    span.set_attribute(key, value)
            try:
                yield span
                span.set_status(Status(StatusCode.OK))
            except Exception as exc:
                span.set_status(Status(StatusCode.ERROR, str(exc)))
                span.record_exception(exc)
                raise

    def create_security_span(
        self,
        operation: str,
        severity: str = "info",
        mitre_technique: str = "",
        threat_type: str = "",
        **kwargs: Any,
    ) -> Span:
        """
        Create a security-specific span with MITRE ATT&CK context.

        Args:
            operation: Security operation name
            severity: Event severity
            mitre_technique: MITRE ATT&CK technique ID
            threat_type: Detected threat type
            **kwargs: Additional attributes

        Returns:
            Created Span object.
        """
        span = self._tracer.start_span(f"msf.security.{operation}")
        span.set_attribute("msf.operation", operation)
        span.set_attribute("msf.severity", severity)
        if mitre_technique:
            span.set_attribute("mitre.technique", mitre_technique)
        if threat_type:
            span.set_attribute("msf.threat_type", threat_type)
        for key, value in kwargs.items():
            span.set_attribute(f"msf.{key}", str(value))
        return span

    def get_trace_id(self) -> str:
        """Get current trace ID or generate new one."""
        current = trace.get_current_span()
        if current and current.get_span_context().trace_id:
            return format(current.get_span_context().trace_id, "032x")
        return secrets.token_hex(16)


_global_telemetry: TelemetryManager | None = None
_telemetry_lock = __import__("threading").Lock()


def get_telemetry() -> TelemetryManager:
    """
    Get the global telemetry manager.

    Returns:
        Global TelemetryManager instance.
    """
    global _global_telemetry
    if _global_telemetry is None:
        with _telemetry_lock:
            if _global_telemetry is None:
                _global_telemetry = TelemetryManager()
    return _global_telemetry


@contextmanager
def create_span(
    name: str,
    attributes: Optional[dict[str, Any]] = None,
) -> Generator[Span, None, None]:
    """
    Convenience context manager for creating spans.

    Args:
        name: Span name
        attributes: Optional attributes

    Yields:
        Active Span.

    Example:
        >>> with create_span("auth.login") as span:
        ...     span.set_attribute("user.id", "u123")
    """
    telemetry = get_telemetry()
    with telemetry.start_span(name, attributes) as span:
        yield span
